homeassistant/bootstrap.py

```diff
@@ -890,72 +858,93 @@ async def _async_set_up_integrations(
+    stages: list[tuple[str, set[str], int | None]] = [
         *(
-            (name, domain_group & domains_to_setup, timeout)
+            (name, domain_group, timeout)
             for name, domain_group, timeout in STAGE_0_INTEGRATIONS
         ),
-        ("stage 1", STAGE_1_INTEGRATIONS & domains_to_setup, STAGE_1_TIMEOUT),
+        ("1", STAGE_1_INTEGRATIONS, STAGE_1_TIMEOUT),
+        ("2", domains, STAGE_2_TIMEOUT),
     ]
 
-    _LOGGER.info("Setting up stage 0 and 1")
-    for name, domain_group, timeout in stage_0_and_1_domains:
-        if not domain_group:
+    _LOGGER.info("Setting up stage 0")
+    for name, domain_group, timeout in stages:
+        stage_domains_unfiltered = domain_group & all_domains
+        if not stage_domains_unfiltered:
+            _LOGGER.info("Nothing to set up in stage %s: %s", name, domain_group)
+            continue
+
+        stage_domains = stage_domains_unfiltered - hass.config.components
+        if not stage_domains:
+            _LOGGER.info("Already set up stage %s: %s", name, stage_domains_unfiltered)
             continue
 
-        _LOGGER.info("Setting up %s: %s", name, domain_group)
-        to_be_loaded = domain_group.copy()
-        to_be_loaded.update(
+        stage_dep_domains_unfiltered = {
             dep
-            for domain in domain_group
-            if (integration := integration_cache.get(domain)) is not None
-            for dep in integration.all_dependencies
+            for domain in stage_domains
+            for dep in all_integrations[domain].all_dependencies
+            if dep not in stage_domains
+        }
+        stage_dep_domains = stage_dep_domains_unfiltered - hass.config.components
+
+        stage_all_domains = stage_domains | stage_dep_domains
+        stage_all_integrations = {
+            domain: all_integrations[domain] for domain in stage_all_domains
+        }
+        # Detect all cycles
+        stage_integrations_after_dependencies = (
+            await loader.resolve_integrations_after_dependencies(
+                hass, stage_all_integrations.values(), stage_all_domains
+            )
         )
-        async_set_domains_to_be_loaded(hass, to_be_loaded)
-        stage_2_domains -= to_be_loaded
+        stage_all_domains = set(stage_integrations_after_dependencies)
+        stage_domains &= stage_all_domains
+        stage_dep_domains &= stage_all_domains
+
+        _LOGGER.info(
+            "Setting up stage %s: %s | %s\nDependencies: %s | %s",
+            name,
+            stage_domains,
+            stage_domains_unfiltered - stage_domains,
+            stage_dep_domains,
+            stage_dep_domains_unfiltered - stage_dep_domains,
+        )
+
+        async_set_domains_to_be_loaded(hass, stage_all_domains)
 
         if timeout is None:
-            await _async_setup_multi_components(hass, domain_group, config)
+            await _async_setup_multi_components(hass, stage_all_domains, config)
         else:
```

Comment:
nit (feel free to ignore), could be written with `continue` to avoid the `else` block:
```
        if timeout is None:
            await _async_setup_multi_components(hass, stage_all_domains, config)
            continue
        try:
            async with hass.timeout.async_timeout(timeout, cool_down=COOLDOWN_TIME):
                await _async_setup_multi_components(hass, stage_all_domains, config)
        except TimeoutError:
            _LOGGER.warning(
                "Setup timed out for stage %s waiting on %s - moving forward",
                name,
                hass._active_tasks,  # noqa: SLF001
            )
```